pkg/cvo/metrics.go

```diff
@@ -156,21 +212,50 @@ func RunMetrics(runContext context.Context, shutdownContext context.Context, lis
 		} else {
 			select {
 			case <-runContext.Done(): // clean shutdown
-			case result := <-resultChannel: // crashed before a shutdown was requested
+			case result := <-resultChannel: // crashed before a shutdown was requested or metrics server recreated
+				if restartServer {
+					klog.Info("Creating metrics server with updated TLS configutation.")
+					server = createHttpServer()
+					go startListening(server, tlsConfig, listenAddress, resultChannel)
+					restartServer = false
+					continue
+				}
 				resultChannelCount--
 				if result.error == nil {
 					klog.Infof("Collected metrics %s goroutine.", result.name)
 				} else {
 					klog.Errorf("Collected metrics %s goroutine: %v", result.name, result.error)
 					loopError = result.error
 				}
+			case event := <-watcher.Events:
+				if event.Op != fsnotify.Chmod && event.Op != fsnotify.Remove &&
+					certsChanged(origCertChecksum, origKeyChecksum, certFile, keyFile) {
+
+					// Update file checksums with latest files.
+					origCertChecksum = checksumFile(certFile)
+					origKeyChecksum = checksumFile(keyFile)
+
+					var tlsErr error
+					tlsConfig, tlsErr = makeTLSConfig(certFile, keyFile)
+					if tlsErr == nil {
+						restartServer = true
+						shutdownHttpServer(shutdownContext, server)
+						continue
+					} else {
+						loopError = fmt.Errorf("Failed to create TLS config: %w", err)
+					}
+				} else {
+					continue
+				}
+			case err = <-watcher.Errors:
+				klog.Warningf("Error from metrics server certificate file watcher: %w", err)
```

Comment:
Add:

```go
loopError = err
```

so we return a non-`nil` `error` in this case (after shutting down and collecting the server result)?  We might end up clobbering it with the `Collected metrics %s goroutine...` case above, but in the case where that server shutdown is not failing, we don't want our caller to feel like we've happily completed our work.

Reply:
So I was thinking that having CVO up with no file watcher is better than no CVO? User would just have to manually restart CVO to pick certs like they do now, But, on the other hand, if we can't create a watcher perhaps that's indicative of other issues?

Reply:
Yeah, I think failing to create the watcher would be something where we'd want a crashlooping CVO so the admin could investigate and file a bug with us (and we might even get logs via Insights).  If someone improperly rotates certs and breaks metrics, they're going to at least start getting TargetDown alerts.  And it's not clear when the next fsnotify event is going to come through and trigger another look.

If crashlooping on these errors is too harsh, our usual soft-complaint channel of "fire an alert" is probably not reliable, because the breakage would be to that metrics pipeline.  We might still be able to complain via ClusterVersion conditions, and I'd be ok with that, but it seems like it would be some work to wire up, and I'm optimistically hoping cred-rotation errors are rare enough that crashlooping is a sufficiently elegant UX when they do come up.  I'm fine punting this sort of handling out to follow-up work which we can pick up if Insights and all show us that cred-rotation issues are more common than I expect them to be.

Reply:
Restructured so this results in a graceful shutdown.

Reply:
In 0e6136c727b05509c, `watcher.Errors` feeds `loopError` and proceeds into graceful shutdown,